blocks/fileio/include/gnuradio-4.0/fileio/BasicFileIo.hpp

```diff
@@ -0,0 +1,307 @@
+    std::filesystem::path filePath(fileName);
+    if (!std::filesystem::exists(filePath.parent_path())) {
+        throw gr::exception(fmt::format("path/file '{}' does not exist.", fileName));
+    }
+
+    std::vector<std::filesystem::path> matchingFiles;
+    std::copy_if(std::filesystem::directory_iterator(filePath.parent_path()), std::filesystem::directory_iterator{}, std::back_inserter(matchingFiles), [&](const auto& entry) { return entry.is_regular_file() && entry.path().string().find(filePath.filename().string()) != std::string::npos; });
+
+    std::sort(matchingFiles.begin(), matchingFiles.end());
+    return matchingFiles;
+}
+
+[[nodiscard]] inline std::uintmax_t getFileSize(const std::filesystem::path& filePath) {
+    if (!std::filesystem::exists(filePath)) {
+        throw std::runtime_error(fmt::format("file '{}' does not exist.", filePath.string()));
+    }
+    return std::filesystem::file_size(filePath);
+}
+
+[[maybe_unused]] inline std::vector<std::string> deleteFilesContaining(const std::string& fileName) {
+    std::filesystem::path filePath(fileName);
+    if (!std::filesystem::exists(filePath.parent_path())) {
+        return {};
+    }
+
+    std::vector<std::string> deletedFiles;
+    for (const auto& entry : std::filesystem::directory_iterator(filePath.parent_path())) {
+        if (entry.is_regular_file() && entry.path().string().find(filePath.filename().string()) != std::string::npos) {
+            deletedFiles.push_back(entry.path().string());
+            std::filesystem::remove(entry.path());
+        }
+    }
+
+    return deletedFiles;
+}
+
+} // namespace detail
+
+enum class FileMode { override, append, multi };
+
+template<typename T>
+struct BasicFileSink : public gr::Block<BasicFileSink<T>> {
+    using Description = Doc<R""(A sink block for writing a stream to a binary file.
+The file can be played back using a 'BasicFileSource' or read by any program that supports binary files (e.g. Python, C, C++, MATLAB).
+For complex types, the binary file contains [float, double]s in IQIQIQ order. No metadata is included with the binary data.)
+Important: this implementation assumes a host-order, CPU architecture specific byte order!)"">;
+    template<typename U, gr::meta::fixed_string description = "", typename... Arguments>
+    using A = gr::Annotated<U, description, Arguments...>; // optional shortening
+
+    PortIn<T> in;
+
+    A<std::string, "file name", Doc<"base filename, prefixed if ">, Visible>              file_name;
+    FileMode                                                                              _mode              = FileMode::override;
+    A<std::string, "mode", Doc<"mode: \"override\", \"append\", \"multi\"">, Visible>     mode               = std::string(magic_enum::enum_name(_mode));
+    A<gr::Size_t, "max bytes per file", Doc<"max bytes per file, 0: infinite ">, Visible> max_bytes_per_file = 0U;
```

Comment:
Maybe it's good to enforce that `max_bytes_per_file % sizeof(T) == 0`. Otherwise it seems that awkward things will happen, as items will be split over different files and lose alignment.

Reply:
The alignment is respected in any case because the counting is done in samples i.e. `sizeof(T)`. The length is more of a maximum size. I'd expect something like in the range of a few MBs and/or GBs.

Reply:
I disagree the `max_bytes_per_file` config parameter is simple and does what its name implies w/o surprises:

The goal of this parameter is to limit the upper bound size of the file in terms of what is relevant for the file system w/o doing or requiring mental math and how many samples this corresponds to by the user. 

The latter can be also quite difficult (and/or independent) depending on the type and `sizeof(T)` since `T` can be not just fixed-sized fundamental types, but more complex ones, and compound types (eventually `Packet<T>`, `Tensor<T>`, or `DataSet<T>`) that are even variably sized.

The use-case example is: "user wants to write data to file and the file not exceeding, for example, 1GB per file regardless whether it stores, `std::int16_t`, `float`, `std::complex<float>`, or a `DataSet<T>`...

This also makes the `max_bytes_per_file % sizeof(T) != 0` difficult to compute and annoying to the user, who usually does not care how big the individual samples/entities in the file are.

*N.B. `_totalBytesWritten` (not `totalBytesWritten` as mentioned) is declared as a private implementation-specific variable and not a public API. We could make it a runtime parameter (then: `total_bytes_written`), for example, for the UI and/or another monitoring process to track how many bytes are written. Together with `max_bytes_per_file` this could provide an indicator of whether progress is made on the block and/or when the file is complete.*